### PR TITLE
Azure Key Vault integration for storage of Tessera public/private key pairs

### PR DESCRIPTION
**Pull request by chris-j-h**
#### Added features 
1. Option to save newly generated keys to a preconfigured Azure Key Vault
1. Retrieve keys from an Azure Key Vault so that they do not have to be stored locally

#### Main areas of interest
- `key-vault` modules: Connecting to the Azure service and providing the means to store and retrieve keys
- `config` modules: Adding further config items and corresponding validation
- `key-generation` module: Storing keys in vault after generation

### CODE REVIEW

File: key-generation/src/main/java/com/quorum/tessera/key/generation/AzureVaultKeyGenerator.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+        final StringBuilder publicId = new StringBuilder();
+        final StringBuilder privateId = new StringBuilder();
+
+        if(filename != null) {
+            final Path path = Paths.get(filename);
+            final String keyVaultId = path.getFileName().toString();
+
+            if(!keyVaultId.matches("^[0-9a-zA-Z\\-]*$")) {
+                throw new RuntimeException("Generated key ID for Azure Key Vault can contain only 0-9, a-z, A-Z and - characters");
```
Comment (melowe) [inline key-generation/src/main/java/com/quorum/tessera/key/generation/AzureVaultKeyGenerator.java]: Can we have a more specialised exception please.

File: key-generation/src/main/java/com/quorum/tessera/key/generation/AzureVaultKeyGenerator.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+        saveKeyInVault(publicId.toString(), keys.getPublicKey());
+        saveKeyInVault(privateId.toString(), keys.getPrivateKey());
+
+        return new AzureVaultKeyPair(publicId.toString(), privateId.toString());
+    }
+
+    private void saveKeyInVault(String id, Key key) {
+        keyVaultService.setSecret(id, key.toString());
+        LOGGER.debug("Key {} saved to vault with id {}", key.toString(), id);
```
Comment (melowe) [inline key-generation/src/main/java/com/quorum/tessera/key/generation/AzureVaultKeyGenerator.java]: Please don't let me see toString() on a possibly null object again for the logger param there is no need toString will be invoked if obj is not null. java.util.Objects.toString(o) its a null safe way of doing this

File: key-generation/src/test/java/com/quorum/tessera/key/generation/AzureVaultKeyGeneratorTest.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    @Test
+    public void keysSavedInVaultWithProvidedVaultId() {
+        final String vaultId = "vaultId";
+        final String pubVaultId = vaultId + "Pub";
+        final String privVaultId = vaultId + "Key";
+
+        final AzureVaultKeyPair result = azureVaultKeyGenerator.generate(vaultId, null);
+
+        verify(keyVaultService, times(2)).setSecret(any(String.class), any(String.class));
+        verify(keyVaultService, times(1)).setSecret(vaultId + "Pub", pub.toString());
```
Comment (melowe) [inline key-generation/src/test/java/com/quorum/tessera/key/generation/AzureVaultKeyGeneratorTest.java]: Please don't depend on the toString method on the Key objects. The fact that this has been used as a business method is an issue that I've addressed in another PR. As a general point toString should only ever be used to provide a readable representation of an instance not to return an encoded value or something functional.

File: key-generation/src/test/java/com/quorum/tessera/key/generation/AzureVaultKeyGeneratorTest.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    @Test
+    public void vaultIdIsFinalComponentOfFilePath() {
+        final String vaultId = "vaultId";
+        final String path = "/some/path/" + vaultId;
+
+        azureVaultKeyGenerator.generate(path, null);
+
+        verify(keyVaultService, times(1)).setSecret(vaultId + "Pub", pub.toString());
```
Comment (melowe) [inline key-generation/src/test/java/com/quorum/tessera/key/generation/AzureVaultKeyGeneratorTest.java]: Key.toString issue again

File: key-generation/src/test/java/com/quorum/tessera/key/generation/KeyGeneratorFactoryTest.java [hunk 5]
```diff
@@ -1,19 +1,26 @@
...
-    public void keyGeneratorIsntNull() {
-
-        final KeyGenerator keyGenerator = KeyGeneratorFactory.newFactory().create();
+    public void vaultKeyGeneratorWhenKeyVaultConfigProvided() {
+        final KeyVaultConfig keyVaultConfig = new KeyVaultConfig("url");
+        final KeyGenerator keyGenerator = KeyGeneratorFactory.newFactory().create(keyVaultConfig);
 
         assertThat(keyGenerator).isNotNull();
-
+        assertThat(keyGenerator.getClass()).isEqualTo(AzureVaultKeyGenerator.class);
```
Comment (melowe) [inline key-generation/src/test/java/com/quorum/tessera/key/generation/KeyGeneratorFactoryTest.java]: assertThat(o).isExactlyInstanceOf(type)

File: key-generation/src/test/java/com/quorum/tessera/key/generation/KeyGeneratorFactoryTest.java [hunk 6]
```diff
@@ -1,19 +1,26 @@
...
     }
 
+    @Test
+    public void fileKeyGeneratorWhenKeyVaultConfigNotProvided() {
+        final KeyGenerator keyGenerator = KeyGeneratorFactory.newFactory().create(null);
 
+        assertThat(keyGenerator).isNotNull();
+        assertThat(keyGenerator.getClass()).isEqualTo(FileKeyGenerator.class);
```
Comment (melowe) [inline key-generation/src/test/java/com/quorum/tessera/key/generation/KeyGeneratorFactoryTest.java]: assertThat(o).isExactlyInstanceOf(type)

File: key-vault/azure-key-vault/pom.xml [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+        <dependency>
+            <groupId>com.microsoft.azure</groupId>
+            <artifactId>adal4j</artifactId>
+            <version>1.6.0</version>
+        </dependency>
+        <dependency>
+            <groupId>com.quorum.tessera</groupId>
+            <artifactId>key-vault-api</artifactId>
+            <version>0.7-SNAPSHOT</version>
```
Comment (melowe) [inline key-vault/azure-key-vault/pom.xml]: version should be defined in parent/root pom in dependencies management section.

File: key-vault/azure-key-vault/src/main/java/com/quorum/tessera/key/vault/AzureKeyVaultClientDelegate.java [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+import com.microsoft.azure.keyvault.KeyVaultClient;
+import com.microsoft.azure.keyvault.models.SecretBundle;
+import com.microsoft.azure.keyvault.requests.SetSecretRequest;
+
+public class AzureKeyVaultClientDelegate {
+    private final KeyVaultClient keyVaultClient;
+
+    public AzureKeyVaultClientDelegate(KeyVaultClient keyVaultClient) {
+        this.keyVaultClient = keyVaultClient;
```
Comment (melowe) [inline key-vault/azure-key-vault/src/main/java/com/quorum/tessera/key/vault/AzureKeyVaultClientDelegate.java]: Objects.requitreNonNull please.

File: key-vault/azure-key-vault/src/main/java/com/quorum/tessera/key/vault/AzureKeyVaultClientFactory.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+ * using adal.
+ */
+public class AzureKeyVaultClientFactory {
+
+    private final ServiceClientCredentials serviceClientCredentials;
+
+    private ExecutorService executorService;
+
+    public AzureKeyVaultClientFactory(ExecutorService executorService) {
+        String clientId = System.getenv("AZURE_CLIENT_ID");
```
Comment (melowe) [inline key-vault/azure-key-vault/src/main/java/com/quorum/tessera/key/vault/AzureKeyVaultClientFactory.java]: Can we avoid System.getenv() assuming this object will be wired in can we access the environment vars in the spring config.

File: key-vault/azure-key-vault/src/main/java/com/quorum/tessera/key/vault/AzureKeyVaultService.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+        }
+
+        this.azureKeyVaultClientDelegate = azureKeyVaultClientDelegate;
+    }
+
+    public String getSecret(String secretName) {
+        SecretBundle secretBundle = azureKeyVaultClientDelegate.getSecret(vaultUrl, secretName);
+
+        if(secretBundle == null) {
+            throw new RuntimeException("Azure Key Vault secret " + secretName + " was not found in vault " + vaultUrl);
```
Comment (melowe) [inline key-vault/azure-key-vault/src/main/java/com/quorum/tessera/key/vault/AzureKeyVaultService.java]: Specialised exception please RuntimeException is too generalised.

**Review by melowe — CHANGES_REQUESTED** [review]

File: tessera-core/src/test/java/com/quorum/tessera/key/KeyPairFactoryTest.java [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+import org.junit.Test;
+
+import java.util.Base64;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.*;
+
+public class KeyPairFactoryTest {
+    KeyPairFactory test;
```
Comment (melowe) [inline tessera-core/src/test/java/com/quorum/tessera/key/KeyPairFactoryTest.java]: private modifiers please. KeyPairFactory is an instance variable named "test", it doesn't seem to be a test

**Review by melowe — CHANGES_REQUESTED** [review]

File: key-generation/src/main/java/com/quorum/tessera/key/generation/AzureVaultKeyGenerator.java (hunk 1 above)
Comment (chris-j-h) [reply key-generation/src/main/java/com/quorum/tessera/key/generation/AzureVaultKeyGenerator.java]: Changed to UnsupportedCharsetException

File: key-vault/azure-key-vault/src/main/java/com/quorum/tessera/key/vault/AzureKeyVaultClientFactory.java (hunk 9 above)
Comment (chris-j-h) [reply key-vault/azure-key-vault/src/main/java/com/quorum/tessera/key/vault/AzureKeyVaultClientFactory.java]: Env vars are now either accessed by Spring or by Provider to allow for easier testing

File: key-vault/azure-key-vault/src/main/java/com/quorum/tessera/key/vault/AzureKeyVaultService.java (hunk 10 above)
Comment (chris-j-h) [reply key-vault/azure-key-vault/src/main/java/com/quorum/tessera/key/vault/AzureKeyVaultService.java]: Custom VaultSecretNotFoundException used

**Review by melowe — APPROVED** [review]

**Review by prd-fox — APPROVED** [review]

File: key-generation/src/main/java/com/quorum/tessera/key/generation/AzureVaultKeyGenerator.java (hunk 2 above)
Comment (chris-j-h) [reply key-generation/src/main/java/com/quorum/tessera/key/generation/AzureVaultKeyGenerator.java]: Noted, toString is no longer used here

File: key-generation/src/test/java/com/quorum/tessera/key/generation/AzureVaultKeyGeneratorTest.java (hunk 3 above)
Comment (chris-j-h) [reply key-generation/src/test/java/com/quorum/tessera/key/generation/AzureVaultKeyGeneratorTest.java]: Now using the encodeToBase64 method

File: key-generation/src/test/java/com/quorum/tessera/key/generation/AzureVaultKeyGeneratorTest.java (hunk 4 above)
Comment (chris-j-h) [reply key-generation/src/test/java/com/quorum/tessera/key/generation/AzureVaultKeyGeneratorTest.java]: Now using the encodeToBase64 method

File: key-generation/src/test/java/com/quorum/tessera/key/generation/KeyGeneratorFactoryTest.java (hunk 5 above)
Comment (chris-j-h) [reply key-generation/src/test/java/com/quorum/tessera/key/generation/KeyGeneratorFactoryTest.java]: Updated

File: key-generation/src/test/java/com/quorum/tessera/key/generation/KeyGeneratorFactoryTest.java (hunk 6 above)
Comment (chris-j-h) [reply key-generation/src/test/java/com/quorum/tessera/key/generation/KeyGeneratorFactoryTest.java]: Updated

File: key-vault/azure-key-vault/pom.xml (hunk 7 above)
Comment (chris-j-h) [reply key-vault/azure-key-vault/pom.xml]: Updated

File: key-vault/azure-key-vault/src/main/java/com/quorum/tessera/key/vault/AzureKeyVaultClientDelegate.java (hunk 8 above)
Comment (chris-j-h) [reply key-vault/azure-key-vault/src/main/java/com/quorum/tessera/key/vault/AzureKeyVaultClientDelegate.java]: Updated

File: tessera-core/src/test/java/com/quorum/tessera/key/KeyPairFactoryTest.java (hunk 11 above)
Comment (chris-j-h) [reply tessera-core/src/test/java/com/quorum/tessera/key/KeyPairFactoryTest.java]: This class has since been removed

**Pull request merged**